### ISSUE TITLE
Simplify TypeDB Clients to return JSON

### ISSUE THREAD

**Issue by richard-julien** [vaticle/typedb#5577]
Dont really know if all drivers have the same behavior today so i create the feature request here instead of the javascript driver.

## Problem to Solve

Getting data easily usable inside nodejs is difficult. If its possible have a direct expression from graql query to JSON answers will be incredible. (See https://rethinkdb.com/ as a source of inspiration)
Keeping a cursor to be able to stream if the result is a list need to be kept, but inside this list I dont think there is a good reason/value to also have cursor instead of just the data.

## Proposed Solution

If I try to imagine something on a basic query

`match $client isa User, has activated true;
            $roles(client: $client, position: $role) isa user_role; $role has name $name;
            $capabilities(position: $role, capability: $capability) isa role_capability; 
            get;`

Expecting this kind of result.

```
Streamed result list
[
  { 
    client: {
      iid: "V42424",
      activated:true,
      roles: [
        { 
          iid: "V21421",
          role: { 
            iid: "V31421",
            name: "Administrator",
            capabilities: [
              {  
                iid: "V60041",
                capability: { iid: "V98941" } 
              },
              {
                iid: "V60042",
                capability: { iid: "V98942" } 
              },
              {
                iid: "V60043",
                capability: { iid: "V98943" } 
              },
            ]
          }
        }
      ]
    }
  },
  {...},
  {...},
]
```
I dont know if its kind of hierarchical result makes sens for you :)

**Comment by thomaschristopherking** [comment]
Hi, just adding that this would be really useful for us as well for two reasons. 1) We often use Neo4j's tools to export to CSV for less tech-savvy users, and this could be a step to having that into workbase. 2) I often use the .get_has(...) call in Python to get all properties of a thing, and then I convert it to a dictionary and eventually a class object. It'd be cleaner to just have JSON returned in the first place.

**Comment by haikalpribadi** [comment]
Yeah I've been thinking of adding something like this to TypeDB Clients. I'll add this to the roadmap.

However, can you help me clarify 2 things, @richard-julien : 

1) Can you help me clarify this sentence?
> Keeping a cursor to be able to stream if the result is a list need to be kept, but inside this list I dont think there is a good reason/value to also have cursor instead of just the data.

2) Just to make sure, with the example query above, you expect `each map` to contain `one answer, right? And based on the exact query above, the `map` would contain each variable (because you did not filter with `get`) as these keys/objects, right?
```
{
  client: { ... }, 
  roles: { ... }, 
  role: { ... }, 
  name: { ... }, 
  capabilities: { ... }, 
  capability: { ... }, 
}
```

**Comment by james-whiteside** [comment]
Worth checking out, apparently this implements a JSON interface for TypeDB:
https://github.com/Blitzapps/blitz-orm

**Comment by flyingsilverfin** [comment]
A very simple first step towards full #6675 stepping stones, would be to just have a direct conversion of ConceptMap to JSON in each client.

For example:
```
{
  "x": { type: "person" },
  "a": { type: "age", valuetype: long, value: 10 },
  "r": { type: friendship },
  "y": { type: "person" }
}
```

**Comment by james-whiteside** [comment]
> A very simple first step towards full #6675 stepping stones, would be to just have a direct conversion of ConceptMap to JSON in each client.
> 
> For example:
> 
> ```
> {
>   "x": { type: "person" },
>   "a": { type: "age", valuetype: long, value: 10 },
>   "r": { type: friendship },
>   "y": { type: "person" }
> }
> ```

I think this is actually a good approach in general. It provides everything the Concept API does so users can avoid ever using it if they don't want to, it allows users to explore the data and utilise properties that they might not even have thought about, and its pretty easy to transform and filter if needs be.

**Comment by lveillard** [comment]
Hello! 
My two cents here. I feel like ConceptMap is a concept pretty related with Java limitations. Seems like a lot of extra work.

With the rust refactoring, I guess it should be possible to do get a list of nodes and its edges, which could be easily used for different use cases later, included the JSON formatting. (And having the permutations as another optional type of query of the DB)

In https://github.com/Blitzapps/blitz-orm we are forced to query each level individually because once you go deep in your query the permutations become really crazy, for instance this query:

```
{
 entity: "City",
 $filter: [name: 'london', name: 'Paris'],
 $fields: [ 'population','code', {
    path: 'authors', 
    $fields: ['name', 'familyName', {
        path: 'books',
        $fields: ['title']
        }
    }
}
```

cant be done in one single query:
```
match
#------------level 1
$c isa city, has $allCityAtts;
{$allCityAtts isa name;} or {$allCityAtts isa code;}; #closest thing to make optionals when querying multiple optional atts so entities with one single att are also gathered.
#level 1 filters
{$c has name 'paris';} or {$c isa city has name London';};

#------------level 2
$a isa author, has $allAuthorAtts;
$c-a (author:$a, city:$c) isa city-author;
....

#-----------level 3
$b isa book...

```

Generates a huge permutation with answers that mix every query variable queried (the relations city-author, author-book, the attributes of each entity etc...) 

So for instance something that should bring
- Two cities, with their populations and codes
- The 2000 authors linked to those cities and their names (attribute edge and value lines)
- The 8000 books linked to those authors, with their titles 

Brings instead a concept map of cosmical size with all the permutations, taking seconds to be queried and extra seconds to be transformed into a tree.

Also conceptMaps are not intuive, almost every candidate that we had that used the native client was confused by it.

So instead we are forced to query level per level, get the iiids of the next level of the tree, and query them in a second round. Which it sends us to <1 second but generates N roundripts to the database.


Getting just the list of nodes and the list of edges without permutation as a list similar of what we get when doing `match $a isa thing` (but including edge data for attributes) would help us doing all the tree query in one single roundtrip and simplify the complex ConceptMapToTree that we had to build 🤔

**Comment by flyingsilverfin** [comment]
Are you just querying for trees instead of graphs?

**Comment by lveillard** [comment]
> Are you just querying for trees instead of graphs?

What is the difference? Aren't trees just graphs with indexed edges? 

The concept I have of a graph is a list of nodes and a list of edges between those nodes, which is exactly what I would expect as an answer when querying data in a graph database. The output of `$t isa thing:` query is the closest i get to that list of nodes and edges as having one single variable removes the permutations.

Imagine a really simple use case without relations, just querying students and teachers and their names, what I will see in Studio graph is just N teacher nodes linked to their names and M student nodes linked to their names:

For a really basic N=M=2 case, we have the list of nodes when inserting them as follows:
![image](https://user-images.githubusercontent.com/29461076/226110152-3c211e6f-3b30-428c-807c-1fed72424f5b.png)

Its graph version is really simple, just 4 entities connected to their respective names:
![image](https://user-images.githubusercontent.com/29461076/226110219-86bf7459-8231-40ca-aae9-8a7af21d8ede.png)

The output tho generates 16 lines of information:
![image](https://user-images.githubusercontent.com/29461076/226110296-b5308bae-9996-4dc2-9108-ddd4a7eeb3bf.png)
Packing a student, with a teacher, with their respective names in one group that does not exist in the graph. There is no relation between 'Carl' and 'Ann' in the graph yet they are packed together in the output of typeDB.

So've inserted 4 lines but i get 16 lines!

The bigger N & M are, the bigger the combinatorial effect becomes, where for instance N=M=3 gives 6 entities connected to their respective 6 attributes and the log output gives 9 groups of 4 lines ( 36 lines) while i just expect something like this:
![image](https://user-images.githubusercontent.com/29461076/226111031-a4663a2c-2ad4-453c-aad1-d19ef3364498.png)

The more variables we define, the bigger the difference between the graph output and the log output is.

My point is that conversion of conceptMap to a JSON seems like a redundant process where we get:
- List of nodes => conceptMapCombinatorialGroups => List of nodes (structured as a JSON)

Seems like a lot of extra work. Getting JSON structured answers would help a lot, but i feel like having the option just to get the list of nodes, and the list of edges would be faster and have better performance.

**Comment by lveillard** [comment]
Probably the most intuitive output would be to have the most common insert structure, which is having attributes specified together with the nodes they belong to.

For instance, this data:
```
insert
$s1 isa student, has name 'Ann';
$s2 isa student, has name 'Bob';
$t1 isa teacher, has name 'Carl';
$t2 isa teacher, has name 'Diana';
```

queried like this:
```
match
$s isa student, has name $s-name;
$t isa teacher, has name $t-name;
```

Would get in the list version
```
    $s iid 0x826e80028000000000000001 isa teacher, has $s-name "Ann" isa name;
    $s iid 0x826e80028000000000000002 isa teacher, has $s-name "Bob" isa name;
    $t iid 0x826e80028000000000000000 isa teacher, has $t-name "Carl" isa name;
    $t iid 0x826e80018000000000000002 isa student, has $t-name "Diana" isa name;
```
Which is exactly what we observe in the graph output.

**Comment by flyingsilverfin** [comment]
I'm just going to address Trees vs Graphs:
a Tree looks like this:
```
   $x
 /      \
$y    $z
 |        |
$a   $b
```
While a graph looks like this:
```
   $x
   /   \
$y   $z
  \    /
   $a
```

The former is actually the only thing GraphQL can query. In addition, I think GraphQL treats each "branch" as independent (correct me if I'm wrong). As a result, because each path is independent the traversal execution is actually really trivial, and allows the return of nested structures with all matching attributes by doing a one-hop at a time traversal.

The graph-shaped queries can have closures and require keeping the permutation of at worst the entire currently visited graph of concepts, as opposed to the tree-shaped which only needs to keep the last concept to generate the next set of vertices.

Hope that makes sense! This is the reason flattening permutations when writing a graph-shaped query engine is harder compared to a tree-shaped one.

**Comment by thomaschristopherking** [comment]
From our perspective we would only need something much simpler. Rather than some magic that flattens the graph structure into JSON, we would probably just want the results of a group statement to be JSON. What goes into the JSON would be up to the query.

Example: `match $p isa person, has attribute $attr; get $p, $attr; group $p;`

Should give something like:

{
"identifier": ["x"],
"first_name": ["Thomas"]
"email_address": ["thomas@example.com", "tking@example.com"]
},
{...}

Maybe the variable names should somehow be included, and also a representation of "p" such as its iid or key _only_. 

But what I'm saying is that whilst I understand the motivation for the harder problem described above, as a user all I want is something I can more easily copy-paste from Studio into other tools and something that's easier to parse in Python (because we actually produce something like the aforementioned JSON ourselves, but I think that functionality should come with TypeDB).

**Comment by lveillard** [comment]
> Hope that makes sense! This is the reason flattening permutations when writing a graph-shaped query engine is harder compared to a tree-shaped one.

I understand it now, thanks! 
GraphQL can tho query stuff in the structure of the graph that you have designed where two branches point to the same branch. Not directly in its structure but yes by reference, for instance querying 

author - country - authors - books
author - books

When JSONs are used not only as a store of values but also of references, even circular dependencies can happen!

---- 
Regarding what @thomaschristopherking said, I agree, that version as a list of things with their attributes linked to them would make a lot of sense, that would be the best way of getting an output

I really think this structure makes a lot of sense:
```json
[
{   "var": "$s", "iid": "0x826e80028000000000000001", "thingType": "entity", "entity": "student",
    "attributes": {"name": "Ann"}
},
{
    "var": "$s-t","iid": "0x826e80028000000000000005","thingType": "relation","relation": "tutorship",
    "players":
    {
        "students": ["0x826e80028000000000000001","0x826e80028000000000000002"],
        "teacher": ["0x826e80028000000000000000"]
    },
    "attributes":  { "fromYear": 2019, "subjects": ["Maths","French"]  }
},
...
]
```

**Comment by dmitrii-ubskii** [comment]
The simple JSON spec in question doesn't alter the set of answers in any way, it only provides a way to serialize the stream of concept maps.

**Comment by benjaminasmall** [comment]
Sorry for the confusion, @lveillard ; this issue is much simpler in scope. See, for example, https://github.com/vaticle/typedb/issues/5577#issuecomment-1473798632 (above).

The more complete solution you correctly identify is on the longer-term roadmap but is a bit out of scope for us now.  Please see  #6675 . We will get there, but we just ask for your patience.

**Comment by lveillard** [comment]
No worries, my fault, seeing the PR linked to this issue confused me.
Happy to see that a solution is in the roadmap!

Respect to https://github.com/vaticle/typedb/issues/5577#issuecomment-1473798632, i'm being redundant and i'm sorry for it, but i don't feel like this is a first step towards https://github.com/vaticle/typedb/issues/6675 its a change of shape that seems to have no impact on the root requirement for https://github.com/vaticle/typedb/issues/6675 , graphQL or the use cases that we exposed here. 

It might seem like it is the case when the specs only describe 1-node cases like the ones in this PR tho, but is not :S

**Comment by flyingsilverfin** [comment]
You're correct to note that the currently added JSON representation is not related directly to this issue, my apologies for polluting it. It's mostly being added to give users a simpler way to get started than learning about `ConceptMap`. The original issue and discussion here still stands!

**Comment by flyingsilverfin** [comment]
This issue has been addressed with the introduction of TypeQL Fetch queries! Released in 2.25.x, please see: https://github.com/vaticle/typeql/pull/300 and https://github.com/vaticle/typedb/pull/6888
